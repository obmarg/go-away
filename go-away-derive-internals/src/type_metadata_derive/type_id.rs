use proc_macro2::{Literal, TokenStream};
use quote::{quote, ToTokens, TokenStreamExt};

pub enum TypeIdCall<'a> {
    Struct(StructTypeIdCall<'a>),
    Variant(VariantTypeIdCall<'a>),
}

impl<'a> TypeIdCall<'a> {
    pub fn for_struct(ident: &'a syn::Ident, generics: &'a syn::Generics) -> Self {
        TypeIdCall::Struct(StructTypeIdCall {
            ident,
            generics: StaticGenerics(generics),
        })
    }

    pub fn for_variant(
        enum_ident: &'a syn::Ident,
        variant_name: &'a syn::Ident,
        generics: &'a syn::Generics,
    ) -> Self {
        TypeIdCall::Variant(VariantTypeIdCall {
            enum_ident,
            variant_name,
            generics: StaticGenerics(generics),
        })
    }
}

impl<'a> ToTokens for TypeIdCall<'a> {
    fn to_tokens(&self, stream: &mut proc_macro2::TokenStream) {
        match self {
            Self::Struct(s) => s.to_tokens(stream),
            Self::Variant(v) => v.to_tokens(stream),
        }
    }
}

pub struct StructTypeIdCall<'a> {
    ident: &'a syn::Ident,
    generics: StaticGenerics<'a>,
}

impl<'a> ToTokens for StructTypeIdCall<'a> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let ident = self.ident;
        let generics = &self.generics;

        tokens.append_all(quote! {
            ::go_away::TypeId::for_type::<#ident #generics>()
        })
    }
}

pub struct VariantTypeIdCall<'a> {
    enum_ident: &'a syn::Ident,
    variant_name: &'a syn::Ident,
    generics: StaticGenerics<'a>,
}

impl<'a> ToTokens for VariantTypeIdCall<'a> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let enum_ident = self.enum_ident;
        let variant_name = Literal::string(&self.variant_name.to_string());
        let generics = &self.generics;

        tokens.append_all(quote! {
            ::go_away::TypeId::for_variant::<#enum_ident #generics, _>(#variant_name)
        })
    }
}

/// Some generics, but with all the lifetimes replaced with 'static.
///
/// We need this to call `TypeId::of` because it doesn't like non static
/// lifetimes
struct StaticGenerics<'a>(&'a syn::Generics);

impl<'a> quote::ToTokens for StaticGenerics<'a> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use syn::{GenericParam, Token};

        if self.0.params.is_empty() {
            return;
        }

        TokensOrDefault(&self.0.lt_token).to_tokens(tokens);

        // Print lifetimes before types and consts, regardless of their
        // order in self.params.
        //
        // TODO: ordering rules for const parameters vs type parameters have
        // not been settled yet. https://github.com/rust-lang/rust/issues/44580
        let mut trailing_or_empty = true;
        for param in self.0.params.pairs() {
            if let GenericParam::Lifetime(_) = *param.value() {
                // Hard code this lifetime to 'static
                tokens.append_all(quote! { 'static });
                param.punct().to_tokens(tokens);
                trailing_or_empty = param.punct().is_some();
            }
        }
        for param in self.0.params.pairs() {
            if let GenericParam::Lifetime(_) = **param.value() {
                continue;
            }
            if !trailing_or_empty {
                <Token![,]>::default().to_tokens(tokens);
                trailing_or_empty = true;
            }
            match *param.value() {
                GenericParam::Lifetime(_) => unreachable!(),
                GenericParam::Type(param) => {
                    // Leave off the type parameter defaults
                    param.ident.to_tokens(tokens);
                }
                GenericParam::Const(param) => {
                    // Leave off the const parameter defaults
                    param.ident.to_tokens(tokens);
                }
            }
            param.punct().to_tokens(tokens);
        }

        TokensOrDefault(&self.0.gt_token).to_tokens(tokens);
    }
}

pub struct TokensOrDefault<'a, T: 'a>(pub &'a Option<T>);

impl<'a, T> quote::ToTokens for TokensOrDefault<'a, T>
where
    T: quote::ToTokens + Default,
{
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self.0 {
            Some(t) => t.to_tokens(tokens),
            None => T::default().to_tokens(tokens),
        }
    }
}
