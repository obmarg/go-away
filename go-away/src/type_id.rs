pub use std::borrow::Cow;

/// A type identifier - used to deduplicate types in the output
#[derive(Hash, Debug, PartialEq, Eq, Clone)]
pub struct TypeId(TypeIdInner);

impl TypeId {
    /// Construct a `TypeId` for a given rust type
    pub fn for_type<T: 'static + ?Sized>() -> Self {
        TypeId(TypeIdInner::Type(std::any::TypeId::of::<T>()))
    }

    /// Construct a `TypeId` for a variant of a rust enum.
    ///
    /// This needs specific support beacuse our output needs types that
    /// don't exist directly in rust.
    pub fn for_variant<T, S>(variant_name: S) -> Self
    where
        T: 'static + ?Sized,
        S: Into<Cow<'static, str>>,
    {
        TypeId(TypeIdInner::Variant {
            parent_enum: std::any::TypeId::of::<T>(),
            variant_name: variant_name.into(),
        })
    }
}

#[derive(Hash, Debug, PartialEq, Eq, Clone)]
pub(super) enum TypeIdInner {
    Type(std::any::TypeId),

    Variant {
        parent_enum: std::any::TypeId,
        variant_name: Cow<'static, str>,
    },
}
