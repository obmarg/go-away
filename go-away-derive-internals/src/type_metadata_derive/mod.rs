use proc_macro2::{Literal, TokenStream};
use quote::quote;
use serde_derive_internals::{
    ast::{Container, Data, Field, Style},
    attr::TagType,
    Ctxt,
};

mod type_id;

use type_id::TypeIdCall;

pub fn type_metadata_derive(ast: &syn::DeriveInput) -> Result<TokenStream, syn::Error> {
    use quote::TokenStreamExt;

    let ctx = Ctxt::new();

    let container =
        Container::from_ast(&ctx, ast, serde_derive_internals::Derive::Deserialize).unwrap();

    match ctx.check() {
        Ok(_) => {}
        Err(errors) => {
            let mut rv = TokenStream::new();
            for err in errors {
                rv.extend(err.to_compile_error());
            }

            return Ok(rv);
        }
    }
    let type_id = TypeIdCall::for_struct(&container.ident, container.generics);

    let ident = &container.ident;
    let name_literal = Literal::string(&ident.to_string());
    let mut inner = TokenStream::new();
    match container.data {
        Data::Enum(variants) if variants.iter().all(|v| matches!(v.style, Style::Unit)) => {
            inner.append_all(quote! {
                let mut rv = types::Enum {
                    name: #name_literal.into(),
                    variants: vec![]
                };
            });
            for variant in variants {
                let variant_name = Literal::string(&variant.ident.to_string());
                let serialized_name = Literal::string(&variant.attrs.name().serialize_name());
                inner.append_all(quote! {
                    rv.variants.push(types::EnumVariant {
                        name: #variant_name.into(),
                        serialized_name: #serialized_name.into(),
                    });
                })
            }

            inner.append_all(quote! {
                FieldType::Named(registry.register_enum(#type_id, rv))
            });
        }
        Data::Enum(variants) if variants.iter().all(|v| matches!(v.style, Style::Newtype)) => {
            let repr = tag_to_representation(container.attrs.tag());
            inner.append_all(quote! {
                let mut rv = types::Union {
                    name: #name_literal.into(),
                    representation: #repr,
                    variants: vec![]
                };
            });
            for variant in variants {
                let variant_name = Literal::string(&variant.ident.to_string());
                let serialized_name = Literal::string(&variant.attrs.name().serialize_name());
                let metadata_call = metadata_call(variant.fields.first().unwrap().ty);
                inner.append_all(quote! {
                    rv.variants.push(
                        types::UnionVariant {
                            name: Some(#variant_name.to_string()),
                            ty: #metadata_call,
                            serialized_name: #serialized_name.to_string()
                        }
                    );
                })
            }
            inner.append_all(quote! {
                FieldType::Named(registry.register_union(#type_id, rv))
            })
        }
        Data::Enum(variants) => {
            let repr = tag_to_representation(container.attrs.tag());
            inner.append_all(quote! {
                let mut rv = types::Union {
                    name: #name_literal.into(),
                    representation: #repr,
                    variants: vec![]
                };
            });
            for variant in variants {
                let variant_name = Literal::string(&variant.ident.to_string());
                let serialized_name = Literal::string(&variant.attrs.name().serialize_name());
                let type_id =
                    TypeIdCall::for_variant(&container.ident, &variant.ident, container.generics);
                let inner_type_block =
                    struct_block(&variant.ident.to_string(), &variant.fields, type_id);
                inner.append_all(quote! {
                    rv.variants.push(
                        types::UnionVariant {
                            name: Some(#variant_name.to_string()),
                            ty: FieldType::Named({#inner_type_block}),
                            serialized_name: #serialized_name.to_string()
                        }
                    );
                })
            }
            inner.append_all(quote! {
                FieldType::Named(registry.register_union(#type_id, rv))
            })
        }
        Data::Struct(Style::Newtype, fields) => {
            let metadata_call = metadata_call(fields.first().unwrap().ty);
            inner.append_all(quote! {
                let nt = types::NewType {
                    name: #name_literal.to_string(),
                    inner: #metadata_call,
                };
                FieldType::Named(registry.register_newtype(#type_id, nt))
            });
        }
        Data::Struct(_, fields) => {
            let struct_block_contents = struct_block(&ident.to_string(), &fields, type_id);
            inner.append_all(quote! {
                let type_ref = {
                    #struct_block_contents
                };
                FieldType::Named(type_ref)
            });
        }
    }

    let (impl_generics, ty_generics, where_clause) = container.generics.split_for_impl();
    Ok(quote! {
        #[automatically_derived]
        impl #impl_generics ::go_away::TypeMetadata for #ident #ty_generics #where_clause {
            fn metadata(registry: &mut ::go_away::TypeRegistry) -> ::go_away::types::FieldType {
                use ::go_away::types::{self, FieldType};
                #inner
            }
        }
    })
}

fn struct_block(name: &str, fields: &[Field], type_id: TypeIdCall<'_>) -> TokenStream {
    use quote::TokenStreamExt;

    let mut rv = TokenStream::new();

    let name_literal = Literal::string(name);
    rv.append_all(quote! {
        let mut st = types::Struct {
            name: #name_literal.into(),
            fields: vec![]
        };
    });
    for field in fields {
        let field_name = name_of_member(&field.member);
        let serialized_name = Literal::string(&field.attrs.name().serialize_name());
        let ty_def = metadata_call(field.ty);
        rv.append_all(quote! {
            st.fields.push(
                types::Field {
                    name: #field_name.into(),
                    serialized_name: #serialized_name.into(),
                    ty: #ty_def
                }
            );
        });
    }
    rv.append_all(quote! {
        registry.register_struct(#type_id, st)
    });

    rv
}

fn tag_to_representation(tag: &TagType) -> proc_macro2::TokenStream {
    match tag {
        TagType::Adjacent { tag, content } => {
            let tag = Literal::string(tag);
            let content = Literal::string(content);
            quote! {
                types::UnionRepresentation::AdjacentlyTagged {
                    tag: #tag.into(),
                    content: #content.into()
                }
            }
        }
        TagType::External => {
            quote! { types::UnionRepresentation::ExternallyTagged }
        }
        TagType::Internal { tag } => {
            let tag = Literal::string(tag);
            quote! {
                types::UnionRepresentation::InternallyTagged {
                    tag: #tag.into()
                }
            }
        }
        TagType::None => {
            quote! { types::UnionRepresentaiton::Untagged }
        }
    }
}

fn metadata_call(ty: &syn::Type) -> proc_macro2::TokenStream {
    match ty {
        syn::Type::Reference(r) => metadata_call(r.elem.as_ref()),
        syn::Type::Paren(r) => metadata_call(r.elem.as_ref()),
        other => quote! { <#other as ::go_away::TypeMetadata>::metadata(registry) },
    }
}

fn name_of_member(member: &syn::Member) -> proc_macro2::Literal {
    use syn::Index;
    match member {
        syn::Member::Named(ident) => Literal::string(&ident.to_string()),
        syn::Member::Unnamed(Index { index, .. }) => Literal::string(&format!("_{}", index)),
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use quote::quote;

    use super::*;

    #[test]
    fn test_struct() {
        assert_snapshot!(test_conversion(quote! {
            struct MyData {
                field_one: String,
                field_two: String
            }
        }))
    }

    #[test]
    fn test_newtype_struct() {
        assert_snapshot!(test_conversion(quote! {
            struct MyData(String);
        }))
    }

    #[test]
    fn test_struct_with_single_field() {
        assert_snapshot!(test_conversion(quote! {
            struct MyData {
                data: String
            }
        }))
    }

    fn test_conversion(ts: proc_macro2::TokenStream) -> String {
        format_code(
            &type_metadata_derive(&syn::parse2(ts).unwrap())
                .unwrap()
                .to_string(),
        )
    }

    fn format_code(text: &str) -> String {
        xshell::cmd!("rustfmt").stdin(text).read().unwrap()
    }
}
