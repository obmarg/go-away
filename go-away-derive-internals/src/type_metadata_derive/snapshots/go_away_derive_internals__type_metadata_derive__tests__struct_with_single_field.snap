---
source: go-away-derive-internals/src/type_metadata_derive/mod.rs
expression: "test_conversion(quote! { struct MyData { data : String } })"

---
#[automatically_derived]
impl ::go_away::TypeMetadata for MyData {
    fn metadata(registry: &mut ::go_away::TypeRegistry) -> ::go_away::types::FieldType {
        use go_away::types::{self, FieldType};
        let type_ref = {
            let mut st = types::Struct {
                name: "MyData".into(),
                fields: vec![],
            };
            st.fields.push(types::Field {
                name: "data".into(),
                serialized_name: "data".into(),
                ty: <String as ::go_away::TypeMetadata>::metadata(registry),
            });
            registry.register_struct(::go_away::TypeId::for_type::<MyData>(), st)
        };
        FieldType::Named(type_ref)
    }
}
